src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -875,27 +885,61 @@ private bool IsAttributeDeclarationTerminator()
                 || this.IsPossibleAttributeDeclaration(); // start of a new one...
         }
 
-        private AttributeListSyntax ParseAttributeDeclaration()
+        private AttributeListSyntax? TryParseAttributeDeclaration(bool inExpressionContext)
         {
-            if (this.IsIncrementalAndFactoryContextMatches && this.CurrentNodeKind == SyntaxKind.AttributeList)
+            if (this.IsIncrementalAndFactoryContextMatches &&
+                this.CurrentNodeKind == SyntaxKind.AttributeList &&
+                !inExpressionContext)
             {
                 return (AttributeListSyntax)this.EatNode();
             }
 
+            using var resetPoint = GetDisposableResetPoint(resetOnDispose: false);
+
             var openBracket = this.EatToken(SyntaxKind.OpenBracketToken);
 
             // Check for optional location :
             var location = IsSomeWord(this.CurrentToken.Kind) && this.PeekToken(1).Kind == SyntaxKind.ColonToken
                 ? _syntaxFactory.AttributeTargetSpecifier(ConvertToKeyword(this.EatToken()), this.EatToken(SyntaxKind.ColonToken))
                 : null;
 
-            var attributes = _pool.AllocateSeparated<AttributeSyntax>();
-            this.ParseAttributes(attributes);
-            return _syntaxFactory.AttributeList(
-                openBracket,
-                location,
-                _pool.ToListAndFree(attributes),
-                this.EatToken(SyntaxKind.CloseBracketToken));
+            var attributesBuilder = _pool.AllocateSeparated<AttributeSyntax>();
+            this.ParseAttributes(attributesBuilder);
+            var attributes = _pool.ToListAndFree(attributesBuilder);
```

Comment:
>var attributes = _pool.ToListAndFree(attributesBuilder); [](http://example.com/codeflow?start=12&length=56)

Consider moving after `if` below. That is, only allocate the list if this is not a collection literal.

Reply:
that would make sense :)